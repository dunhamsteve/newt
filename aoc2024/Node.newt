../aoc2023/Node.newt