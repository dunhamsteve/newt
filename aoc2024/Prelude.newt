../aoc2023/Prelude.newt