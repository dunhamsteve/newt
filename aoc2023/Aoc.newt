module Aoc

import Prelude

nums : String → List Int
nums s = map stringToInt $ filter (_/=_ "") $ split (trim s) " "

isDigit : Char -> Bool
isDigit '0' = True
isDigit '1' = True
isDigit '2' = True
isDigit '3' = True
isDigit '4' = True
isDigit '5' = True
isDigit '6' = True
isDigit '7' = True
isDigit '8' = True
isDigit '9' = True
isDigit _ = False

indexOf? : ∀ a. {{Eq a}} → a → List a → Maybe Nat
indexOf? {a} z xs = go Z z xs
  where
    go : Nat → a → List a → Maybe Nat
    go ix z Nil = Nothing
    go ix z (x :: xs) =
      if z == x then Just ix else go (S ix) z xs

-- if_then_else shorthand
-- Lean version uses a decidable instead of Bool
ite : ∀ a. Bool → a → a → a
ite c t e = if c then t else e

-- probably not super efficient, but it works
qsort : ∀ a. (a → a → Bool) → List a → List a
qsort lt Nil = Nil
qsort lt (x :: xs) = qsort lt (filter (λ y => not $ lt x y) xs) ++ x :: qsort lt (filter (lt x) xs)
