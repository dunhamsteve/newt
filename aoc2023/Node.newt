module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv)`
pfunc readFile uses (fs MkIORes Unit) : (fn : String) -> IO String := `(fn) => (w) => MkIORes(Unit, fs.readFileSync(fn, 'utf8'), w)`
