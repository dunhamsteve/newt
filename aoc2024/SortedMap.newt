module SortedMap

import Prelude

-- TODO We'll want to replace Ord/Eq with (a → Ordering) (and rewrite most of our aoc solutions...)
-- data Ordering : U where
--   LT EQ GT : Ordering

data T23 : Nat -> U -> U -> U where
  Leaf : ∀ k v. k -> v -> T23 Z k v
  Node2 : ∀ h k v. T23 h k v -> k -> T23 h k v -> T23 (S h) k v
  Node3 : ∀ h k v. T23 h k v -> k -> T23 h k v -> k -> T23 h k v -> T23 (S h) k v

lookupT23 : ∀ h k v. {{Ord k}} {{Eq k}} -> k -> T23 h k v -> Maybe (k × v)
lookupT23 key (Leaf k v)= if k == key then Just (k,v) else Nothing
lookupT23 key (Node2 t1 k1 t2) =
  if key <= k1 then lookupT23 key t1 else lookupT23 key t2
lookupT23 key (Node3 t1 k1 t2 k2 t3) =
  if key <= k1 then lookupT23 key t1
  else if key <= k2 then lookupT23 key t2
  else lookupT23 key t3

insertT23 : ∀ h k v. {{Ord k}} {{Eq k}} -> k -> v -> T23 h k v -> Either (T23 h k v) (T23 h k v × k × T23 h k v)
insertT23 key value (Leaf k v) =
  if key == k then Left (Leaf key value)
  else if key <= k then Right (Leaf key value, key, Leaf k v)
  else Right (Leaf k v, k, Leaf key value)
insertT23 key value (Node2 t1 k1 t2) =
  if key <= k1 then
    case insertT23 key value t1 of
      Left t1' => Left (Node2 t1' k1 t2)
      Right (a,b,c) => Left (Node3 a b c k1 t2)
  else case insertT23 key value t2 of
      Left t2' => Left (Node2 t1 k1 t2')
      Right (a,b,c) => Left (Node3 t1 k1 a b c)
insertT23 key value (Node3 t1 k1 t2 k2 t3) =
  if key <= k1 then
    case insertT23 key value t1 of
      Left t1' => Left (Node3 t1' k1 t2 k2 t3)
      Right (a,b,c) => Right (Node2 a b c, k1, Node2 t2 k2 t3)
  else if key <= k2 then
    case insertT23 key value t2 of
      Left t2' => Left (Node3 t1 k1 t2' k2 t3)
      Right (a,b,c) => Right (Node2 t1 k1 a, b, Node2 c k2 t3)
  else
    case insertT23 key value t3 of
      Left t3' => Left (Node3 t1 k1 t2 k2 t3')
      Right (a,b,c) => Right (Node2 t1 k1 t2, k2, Node2 a b c)

-- This is cribbed from Idris. Deleting nodes takes a bit of code.
Hole : Nat → U → U → U
Hole Z k v = Unit
Hole (S n) k v = T23 n k v

Node4 : ∀ k v h. T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → T23 (S (S h)) k v
Node4 t1 k1 t2 k2 t3 k3 t4 = Node2 (Node2 t1 k1 t2) k2 (Node2 t3 k3 t4)

Node5 : ∀ k v h. T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → T23 (S (S h)) k v
Node5 a b c d e f g h i = Node2 (Node2 a b c) d (Node3 e f g h i)

Node6 : ∀ k v h. T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → T23 (S (S h)) k v
Node6 a b c d e f g h i j k = Node2 (Node3 a b c d e) f (Node3 g h i j k)

Node7 : ∀ k v h. T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → T23 (S (S h)) k v
Node7 a b c d e f g h i j k l m = Node3 (Node3 a b c d e) f (Node2 g h i) j (Node2 k l m)

merge1 : ∀ k v h. T23 h k v → k → T23 (S h) k v → k → T23 (S h) k v → T23 (S (S h)) k v
merge1 a b (Node2 c d e) f (Node2 g h i) = Node5 a b c d e f g h i
merge1 a b (Node2 c d e) f (Node3 g h i j k) = Node6 a b c d e f g h i j k
merge1 a b (Node3 c d e f g) h (Node2 i j k) = Node6 a b c d e f g h i j k
merge1 a b (Node3 c d e f g) h (Node3 i j k l m) = Node7 a b c d e f g h i j k l m

merge2 : ∀ k v h. T23 (S h) k v → k → T23 h k v → k → T23 (S h) k v → T23 (S (S h)) k v
merge2 (Node2 a b c) d e f (Node2 g h i) = Node5 a b c d e f g h i
merge2 (Node2 a b c) d e f (Node3 g h i j k) = Node6 a b c d e f g h i j k
merge2 (Node3 a b c d e) f g h (Node2 i j k) = Node6 a b c d e f g h i j k
merge2 (Node3 a b c d e) f g h (Node3 i j k l m) = Node7 a b c d e f g h i j k l m

merge3 : ∀ k v h. T23 (S h) k v → k → T23 (S h) k v → k → T23 h k v → T23 (S (S h)) k v
merge3 (Node2 a b c) d (Node2 e f g) h i = Node5 a b c d e f g h i
merge3 (Node2 a b c) d (Node3 e f g h i) j k = Node6 a b c d e f g h i j k
merge3 (Node3 a b c d e) f (Node2 g h i) j k = Node6 a b c d e f g h i j k
merge3 (Node3 a b c d e) f (Node3 g h i j k) l m = Node7 a b c d e f g h i j k l m

-- height is erased in the data everywhere but the top, but needed for case
-- I wonder if we could use a 1 + 1 + 1 type instead of Either Tree Hole and condense this
deleteT23 : ∀ k v. {{Ord k}} {{Eq k}} → (h : Nat) -> k -> T23 h k v -> Either (T23 h k v) (Hole h k v)
deleteT23 Z key (Leaf k v) = if k == key then Right MkUnit else Left (Leaf k v)
deleteT23 (S Z) key (Node2 t1 k1 t2) =
  if key <= k1
  then case deleteT23 Z key t1 of
    Left t1 => Left (Node2 t1 k1 t2)
    Right _ => Right t2
  else case deleteT23 Z key t2 of
    Left t2 => Left (Node2 t1 k1 t2)
    Right MkUnit => Right t1
deleteT23 (S Z) key (Node3 t1 k1 t2 k2 t3) =
  if key <= k1
  then case deleteT23 _ key t1 of
    Left t1 => Left (Node3 t1 k1 t2 k2 t3)
    Right MkUnit => Left (Node2 t2 k2 t3)
  else if key <= k2 then case deleteT23 _ key t2 of
    Left t2 => Left (Node3 t1 k1 t2 k2 t3)
    Right _ => Left (Node2 t1 k1 t3)
  else case deleteT23 _ key t3 of
    Left t3 => Left (Node3 t1 k1 t2 k2 t3)
    Right _ => Left (Node2 t1 k1 t2)
deleteT23 (S (S h)) key (Node2 t1 k1 t2) =
  if key <= k1
  then case deleteT23 (S h) key t1 of
    Left t1 => Left (Node2 t1 k1 t2)
    Right t1 => case t2 of
      Node2 t2' k2' t3' => Right (Node3 t1 k1 t2' k2' t3')
      Node3 t2 k2 t3 k3 t4 => Left $ Node4 t1 k1 t2 k2 t3 k3 t4
  else case deleteT23 _ key t2 of
    Left t2 => Left (Node2 t1 k1 t2)
    Right t2 => case t1 of
      Node2 a b c => Right (Node3 a b c k1 t2)
      Node3 a b c d e => Left (Node4 a b c d e k1 t2)
deleteT23 (S (S h)) key (Node3 t1 k1 t2 k2 t3) =
  if key <= k1
  then case deleteT23 _ key t1 of
    Left t1 => Left (Node3 t1 k1 t2 k2 t3)
    Right t1 => Left (merge1 t1 k1 t2 k2 t3)
  else if key <= k2 then case deleteT23 _ key t2 of
    Left t2 => Left (Node3 t1 k1 t2 k2 t3)
    Right t2 => Left (merge2 t1 k1 t2 k2 t3)
  else case deleteT23 _ key t3 of
    Left t3 => Left (Node3 t1 k1 t2 k2 t3)
    Right t3 => Left (merge3 t1 k1 t2 k2 t3)

treeLeft : ∀ h k v. T23 h k v → (k × v)
treeLeft (Leaf k v) = (k, v)
treeLeft (Node2 t1 _ _) = treeLeft t1
treeLeft (Node3 t1 _ _ _ _) = treeLeft t1

treeRight : ∀ h k v. T23 h k v → (k × v)
treeRight (Leaf k v) = (k, v)
treeRight (Node2 _ _ t2) = treeRight t2
treeRight (Node3 _ _ _ _ t3) = treeRight t3


data SortedMap : U -> U -> U where
  EmptyMap : ∀ k v. SortedMap k v
  -- not erased so we know what happens in delete
  MapOf : ∀ k v. {h : Nat} →  T23 h k v -> SortedMap k v

deleteMap : ∀ k v. {{Ord k}} {{Eq k}} → k → SortedMap k v → SortedMap k v
deleteMap key EmptyMap = EmptyMap
-- REVIEW if I split h separately in a nested case, it doesn't sort out Hole
deleteMap key (MapOf {k} {v} {Z} tree) = case deleteT23 Z key tree of
  Left t => MapOf t
  Right t => EmptyMap
deleteMap key (MapOf {k} {v} {S n} tree) = case deleteT23 (S n) key tree of
  Left t => MapOf t
  Right t => MapOf t

leftMost : ∀ k v. SortedMap k v → Maybe (k × v)
leftMost EmptyMap = Nothing
leftMost (MapOf m) = Just (treeLeft m)

rightMost : ∀ k v. SortedMap k v → Maybe (k × v)
rightMost EmptyMap = Nothing
rightMost (MapOf m) = Just (treeRight m)

-- TODO issue with metas and case if written as `do` block
pop : ∀ k v. {{Eq k}} {{Ord k}} → SortedMap k v → Maybe ((k × v) × SortedMap k v)
pop m = case leftMost m of
    Just (k,v) => Just ((k,v), deleteMap k m)
    Nothing => Nothing

lookupMap : ∀ k v. {{Ord k}} {{Eq k}} -> k -> SortedMap k v -> Maybe (k × v)
lookupMap k EmptyMap = Nothing
lookupMap k (MapOf map) = lookupT23 k map

updateMap : ∀ k v. {{Ord k}} {{Eq k}} -> k -> v -> SortedMap k v -> SortedMap k v
updateMap k v EmptyMap = MapOf $ Leaf k v
updateMap k v (MapOf map) = case insertT23 k v map of
  Left map' => MapOf map'
  Right (a, b, c) => MapOf (Node2 a b c)

toList : ∀ k v. SortedMap k v → List (k × v)
toList {k} {v} (MapOf smap) = reverse $ go smap Nil
  where
    go : ∀ h. T23 h k v → List (k × v) → List (k × v)
    go (Leaf k v) acc = (k, v) :: acc
    go (Node2 t1 k1 t2) acc = go t2 (go t1 acc)
    go (Node3 t1 k1 t2 k2 t3) acc = go t3 $ go t2 $ go t1 acc
toList _ = Nil
