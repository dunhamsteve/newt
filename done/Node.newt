module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv.slice(1))`
pfunc readFile uses (fs MkIORes Left Right) : (fn : String) -> IO (Either String String) := `(fn) => (w) => {
  let result
  try {
    let content = fs.readFileSync(fn, 'utf8')
    result = Right(undefined, undefined, content)
  } catch (e) {
    let err = ""+e
    result = Left(undefined, undefined, e)
  }
  return MkIORes(undefined, result, w)
}`

-- I wonder if I should automatically `uses` the constructors in the types
pfunc writeFile uses (fs MkIORes MkUnit) : String → String → IO (Either String Unit) := `(fn, content) => (w) => {
  let result
  try {
    fs.writeFileSync(fn, content, 'utf8')
    result = Right(undefined, undefined, MkUnit)
  } catch (e) {
      let err = ""+e
      result = Left(undefined, undefined, e)
  }
  return MkIORes(undefined, result, w)
}`

-- maybe System.exit or something, like the original putStrLn msg >> exitFailure
pfunc exitFailure : ∀ a. String → a := `(_, msg) => { 
  console.log(msg);
  process.exit(1);
}`

