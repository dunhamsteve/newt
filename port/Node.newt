module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv.slice(1))`
pfunc readFile uses (fs MkIORes Left Right) : (fn : String) -> IO (Either String String) := `(fn) => (w) => {
  let result
  try {
    let content = fs.readFileSync(fn, 'utf8')
    result = Right(null, null, content)
  } catch (e) {
    let err = ""+e
    result = Left(null, null, e)
  }
  return MkIORes(null, result, w)
}`

-- I wonder if I should automatically `uses` the constructors in the types
pfunc writeFile uses (fs MkIORes MkUnit) : String → String → IO (Either String Unit) := `(fn, content) => (w) => {
  let result
  try {
    fs.writeFileSync(fn, content, 'utf8')
    result = Right(null, null, MkUnit)
  } catch (e) {
      let err = ""+e
      result = Left(null, null, e)
  }
  return MkIORes(null, result, w)
}`

-- maybe System.exit or something, like the original putStrLn msg >> exitFailure
pfunc exitFailure : ∀ a. String → a := `(_, msg) => { 
  console.log(msg);
  process.exit(1);
}`

