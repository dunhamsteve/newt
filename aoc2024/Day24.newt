module Day24

import Prelude
import Node
import Aoc
import SortedMap

data Value : U where
  One Zero None : Value

data Op : U where
  Xor And Or : Op

instance Eq Op where
  Xor == Xor = True
  And == And = True
  Or == Or = True
  _ == _ = False

record Gate where
  constructor GT
  in1 : String
  in2 : String
  op : Op
  out : String

Wire : U
Wire = String × Int

parseFile : String → Either String (List Wire × List Gate)
parseFile text = do
  let (a :: b :: Nil) = split (trim text) "\n\n" | _ => Left "too many parts"
  wires <- traverse parseWire $ lines a
  gates <- traverse parseGate $ lines b
  Right (wires, gates)

  where
    parseWire : String → Either String (String × Int)
    parseWire s = let (a :: b :: Nil) = split s ": " | _ => Left $ "bad wire: " ++ s in
      Right (a, stringToInt b)

    getOp : String → Either String Op
    getOp "XOR" = Right Xor
    getOp "AND" = Right And
    getOp "OR"  = Right Or
    getOp op = Left $ "bad op " ++ op

    parseGate : String → Either String Gate
    parseGate s = do
      let (in1 :: op :: in2 :: _ :: out :: Nil) = split s " " | _ => Left $ "bad gate: " ++ s
      op <- getOp op
      Right $ GT in1 in2 op out


State : U
State = SortedMap String Int


mkFire : List Gate → SortedMap String (List String)
mkFire gates = go EmptyMap gates
  where
    go : SortedMap String (List String) → List Gate → SortedMap String (List String)
    go acc Nil = acc
    go acc (g :: gs) =
      let acc = updateMap (in1 g) (out g :: fromMaybe Nil (snd <$> lookupMap (in1 g) acc)) acc in
      let acc = updateMap (in2 g) (out g :: fromMaybe Nil (snd <$> lookupMap (in2 g) acc)) acc in
      go acc gs



exec : SortedMap String Gate → SortedMap String (List String) → List String → State → State
exec gateMap fireMap todo state = go todo state
  where
    combine : Op → Int → Int → Int
    combine And 1 x = x
    combine Xor 1 1 = 0
    combine Xor 1 0 = 1
    combine Xor 0 1 = 1
    combine Or 1 _ = 1
    combine Or _ 1 = 1
    combine _ _ _ = 0

    go : List String → State → State
    go Nil st = st
    go (x :: xs) st = case lookupMap x st of
      -- already done
      Just _ => go xs st
      Nothing => case lookupMap' x gateMap of
        Nothing => go xs st
        Just gate => case (lookupMap' (in1 gate) st, lookupMap' (in2 gate) st) of
          (Just x, Just y) =>
            let v = combine (op gate) x y
                st = updateMap (out gate) v st
                foo = fromMaybe Nil $ lookupMap' (out gate) fireMap
                xs' = foo ++ xs
            in go xs' st
          _ => go xs st

bitsToInt : Int → List Int → Int
bitsToInt _ Nil = 0
bitsToInt m (x :: xs) = m * x + bitsToInt (2 * m) xs


-- so I can find the lowest bit that is wrong
-- or trace out the circuit?

infixl 5 _%_
pfunc _%_ : Int → Int → Int := `(x,y) => x % y`

label : Char -> Int -> String
label c bit = pack $ c :: chr (bit / 10 + 48) :: chr ((bit % 10) + 48) :: Nil

range : Int → Int → List Int
range a b = if a < b then a :: range (a + 1) b else Nil

swapPins : String → String → Gate → Gate
swapPins a g (GT i1 i2 op out) =
  if out == a then GT i1 i2 op g
  else if out == g then GT i1 i2 op a
  else GT i1 i2 op out

fail : ∀ a. String -> a
fail msg = let x = trace "FAIL" msg in ?

check : List Gate → List Int → String → Either (String × String) Unit
check gates Nil carry = Right MkUnit
check gates (bit :: bits) carry = do
  let xl = label 'x' bit
  let (Just a1) = matchIn xl And | _ => fail $ "no a1 " ++ show bit
  let (Just x1) = matchIn xl Xor | _ => fail $ "no x1 " ++ show bit
  -- just peel of the carry for bit0
  let (False) = bit == 0 | _ => check gates bits (out a1)

  let (Just x2) = matchIn carry Xor
    | _ => do
      -- carry is bad
      let (Just x2) = matchOut (label 'z' bit) Xor | _ => fail $ "no x2 for " ++ show bit
      if in1 x2 == out x1 then Left (in2 x2, carry) else Left (in1 x2, carry)
  let (Just a2) = matchIn carry And | _ => fail $ "no a2 " ++ show bit
  checkPins x2 carry (out x1)
  checkPins a2 carry (out x1)

  case (matchIn (out a1) Or, matchIn (out a2) Or) of
    (Nothing, Nothing) => fail "todo2"
    (Just g, Nothing) => checkPins g (out a1) (out a2)
    (Nothing, Just g) => checkPins g (out a1) (out a2)
    (Just g, Just g') =>
      if out g == out g'
        then check gates bits (out g)
        else fail "todo"

  where
    checkPins : Gate → String → String → Either (String × String) Unit
    checkPins g a b =
      if a == in1 g && b == in2 g then pure MkUnit
      else if a == in2 g && b == in1 g then pure MkUnit
      else if a == in1 g then Left (b, in2 g)
      else if a == in2 g then Left (b, in1 g)
      else if b == in1 g then Left (a, in2 g)
      else Left (a, in1 g)

    matchOut : String -> Op -> Maybe Gate
    matchOut l o = find (\ g => out g == l && op g == o) gates

    matchIn : String -> Op -> Maybe Gate
    matchIn l o = find (\ g => (in1 g == l || in2 g == l) && op g == o) gates

trampoline : List Gate → List String → IO (List String)
trampoline gates acc = do
  case check gates (range 0 44) "" of
    Right _ => pure acc
    Left (a,b) => do
      putStrLn $ "SWAP " ++ a ++ " and " ++ b
      trampoline (map (swapPins a b) gates) (a :: b :: acc)

run : String -> IO Unit
run fn = do
  putStrLn fn
  text <- readFile fn
  let (Right (wires, gates)) = parseFile text | Left msg => putStrLn $ "fail: " ++ msg
  let state = foldMap const EmptyMap wires
  let gateMap = foldMap const EmptyMap $ map (\ gate => (out gate, gate)) gates
  let fireMap = mkFire gates
  let init = join $ map snd $ mapMaybe (\ wire => lookupMap wire fireMap) $ map fst wires
  let state' = exec gateMap fireMap init state
  let bits = map snd $ filter (_<_ "z" ∘ fst) $ toList state'
  let p1 = bitsToInt 1 bits
  putStrLn $ "part1 " ++ show p1

  stuff <- trampoline gates Nil
  putStrLn $ "part2 " ++ (joinBy "," $ qsort _<_ stuff)
  putStrLn ""

main : IO Unit
main = do
  -- run "aoc2024/day24/eg.txt"
  -- run "aoc2024/day24/eg2.txt"
  -- this is the only adder circuit
  run "aoc2024/day24/input.txt"
