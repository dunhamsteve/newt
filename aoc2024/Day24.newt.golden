aoc2024/day24/input.txt
part1 51745744348272
SWAP z18 and hmt
SWAP bfq and z27
SWAP z31 and hkh
SWAP fjp and bng
part2 bfq,bng,fjp,hkh,hmt,z18,z27,z31

